### DISCUSSION TITLE
ItemsService queries async/await order

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#10109] [Archive]
Hi there,

I've been staring at this for hours now and I'm missing something but can't seem to figure out what. Check this piece of code:

```js
  this.logger.info(`Publishing supplier product with IDs ${aSupplierProductIdArray}`);
  const { schema } = this;
  this.productVariantItemsService = new this.ItemsService('product_variants', { schema });
  this.supplierProductItemsService = new this.ItemsService('supplier_product_variants', { schema });

  // Get all given items from array
  const supplierProductToPublish = await this.supplierProductItemsService.readMany(aSupplierProductIdArray, {
    fields: ['id', 'icecat_id'],
  });

  await supplierProductToPublish.forEach(async (aSupplierProductId) => {
    // Loop over all items
    this.logger.debug(
      `Supplier product with ID ${aSupplierProductId.id} has icecat id ${aSupplierProductId.icecat_id}`
    );
    // second check if another Item (different collection) exists.
    const filter = { icecat_id: { _eq: supplierProductToPublish.icecat_id } };
    const existingProductVariants = await this.productVariantItemsService.readByQuery({
      fields: ['id'],
      filter,
    });

    this.logger.debug(`${existingProductVariants.length}`);
  });
```

Now what I was thinking/hoping/looking for is this output:
```
✨ Publishing supplier product with IDs 1,2
🐛 Supplier product with ID 1 has icecat id icecatid1234
🐛 0
🐛 Supplier product with ID 2 has icecat id icecatid1234
🐛 0
```

Instead I'm getting this:
```
✨ Publishing supplier product with IDs 1,2
🐛 Supplier product with ID 2 has icecat id icecatid1234
🐛 Supplier product with ID 1 has icecat id icecatid1234
🐛 0
🐛 0
```

Why is this loop not waiting for the existingProductVariants call? And why is ID 2 returned first even? What am I missing?

Cheers!

Kees.

**Comment by azrikahar** [comment]
This is an "issue", or rather a quirk that is technically by design in JavaScript, not something to do with ItemsService or Directus to be clear. Async/await behaves differently in `forEach` vs a typical `for` loop or `for ... of` loop, so you should avoid using forEach to do what you were looking to achieve. 

These resources should offer more insight on this topic:

- https://gist.github.com/joeytwiddle/37d2085425c049629b80956d3c618971

- https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop/67597038#67597038

**Reply by keesvanbemmel** [reply]
Thanks @azrikahar

**Reply by kadlinobit** [reply]
Hi @azrikahar, 
I think I ran into similar issue, but there is no forEach loop in my code. 

I have this hook extension where I'm calling ItemsService: 

```js
export default ({ filter, action }, context) => {

	const { ItemsService } = context.services
	const { ForbiddenException, InvalidPayloadException } = context.exceptions

	filter('user_dj_follow.items.create', async (input, {collection}, {schema, accountability}) => {
		
		const followService = new ItemsService('user_dj_follow', { schema, accountability });
		const filters = {
				"_and" : [
					{"user_created": {"_eq": accountability.user}},
					{"dj": {"_eq": input.dj}}
				]
			}

			const results = await followService.readByQuery({ filters, fields: ['id'] })

			if (results.length > 0) {
				throw new InvalidPayloadException('You cant follow DJ more than once');
			}
			return input
	});
};
```

The problem is that the call `const results = await followService.readByQuery({ filters, fields: ['id'] })` always times out. When I use classical `Promise.then()`, it's working just fine - but I need blocking code here. 

The error I'm getting is:
```
KnexTimeoutError Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call? 
```
Do you have any idea why this could be happening? 

Thank you so much!

**Reply by rijkvanzanten** [reply]
@kadlinobit It's most likely the fact that the `readByQuery` uses a new database connection during the current transaction, which causes the pool to fill by one. If your database has a hard limit on the amount of simultaneous connections, this will stall and fail.

For most databases, this default limit is high enough where you'd never notice it. That being said, SQLite is notorious for only allowing a single connection at a time. Are you using SQLite by chance? 

The way around this is by passing the database connection used in the hook to the service, so it can reuse the same transaction:


```js
filter('user_dj_follow.items.create', async (input, { collection }, { database, schema, accountability }) => {
  const followService = new ItemsService('user_dj_follow', { schema, accountability, knex: database });
```

**Reply by kadlinobit** [reply]
Yes, that was exactly my problem, thank you @rijkvanzanten !
